accumulator/batchproof.go

```diff
@@ -112,122 +112,134 @@ func FromBytesBatchProof(b []byte) (BatchProof, error) {
+// and the subset of roots that was computed.
+func verifyBatchProof(bp BatchProof, roots []Hash, numLeaves uint64,
+	// cached should be a function that fetches nodes from the pollard and indicates whether they
+	// exist or not, this is only useful for the pollard and nil should be passed for the forest.
+	cached func(pos uint64) (bool, Hash)) (bool, [][3]node, []node) {
 	if len(bp.Targets) == 0 {
-		return true, nil
+		return true, nil, nil
 	}
 
-	// Construct a map with positions to hashes
-	proofmap, err := bp.Reconstruct(numLeaves, rows)
-	if err != nil {
-		fmt.Printf("VerifyBlockProof Reconstruct ERROR %s\n", err.Error())
-		return false, proofmap
+	if cached == nil {
+		cached = func(_ uint64) (bool, Hash) { return false, empty }
 	}
 
-	rootPositions, rootRows := getRootsReverse(numLeaves, rows)
-
-	// partial forest is built, go through and hash everything to make sure
-	// you get the right roots
+	rows := treeRows(numLeaves)
+	proofPositions, computablePositions := ProofPositions(bp.Targets, numLeaves, rows)
+	// targetNodes holds nodes that are known, on the bottom row those are the targets,
+	// on the upper rows it holds computed nodes.
+	// rootCandidates holds the roots that where computed, and have to be compared to the actual roots
+	// at the end.
+	targetNodes := make([]node, 0, len(bp.Targets)*int(rows))
+	rootCandidates := make([]node, 0, len(roots))
+	// trees is a slice of 3-Tuples, each tuple represents a parent and its children.
+	// tuple[0] is the parent, tuple[1] is the left child and tuple[2] is the right child.
+	// trees holds the entire proof tree of the batchproof in this way, sorted by the tuple[0].
+	trees := make([][3]node, 0, len(computablePositions))
+	// initialise the targetNodes for row 0.
+	// TODO: this would be more straight forward if bp.Proofs wouldn't contain the targets
+	proofHashes := make([]Hash, 0, len(proofPositions))
+	targets := bp.Targets
+	var targetsMatched uint64
+	for len(targets) > 0 {
+		if targets[0] == numLeaves-1 && numLeaves&1 == 1 {
+			// row 0 root.
+			rootCandidates = append(rootCandidates, node{Val: roots[0], Pos: targets[0]})
+			bp.Proof = bp.Proof[1:]
+			break
+		}
 
-	tagRow := bp.Targets
-	nextRow := []uint64{}
-	sortUint64s(tagRow) // probably don't need to sort
+		if uint64(len(proofPositions)) > targetsMatched &&
+			targets[0]^1 == proofPositions[targetsMatched] {
+			lr := targets[0] & 1
+			targetNodes = append(targetNodes, node{Pos: targets[0], Val: bp.Proof[lr]})
+			proofHashes = append(proofHashes, bp.Proof[lr^1])
+			targetsMatched++
+			bp.Proof = bp.Proof[2:]
+			targets = targets[1:]
+			continue
+		}
 
-	// TODO it's ugly that I keep treating the 0-row as a special case,
-	// and has led to a number of bugs.  It *is* special in a way, in that
-	// the bottom row is the only thing you actually prove and add/delete,
-	// but it'd be nice if it could all be treated uniformly.
+		if len(bp.Proof) < 2 || len(targets) < 2 {
```

Comment:
What would be a case where this clause is true?

Reply:
i added a comment there, let me know if that explains it.